### PR TITLE
Persistent model indices, part 1: Persistent model indices and granular updates

### PR DESCRIPTION
**Pull request by sin-ack**
This is the first in multiple PRs which attempt to fix the biggest issue with LibGUI models which is the fact that they lose state anytime their data updates, even if that change could have been handled granularly. This causes things like the File Manager tree view collapsing when certain operations happen, and items in list/table views losing selection when even the smallest changes happen. The solution is to have a system in place which allows the model to emit a series of operations which will tell the views what has happened to the model's data, so that the views can react without losing their selection/hover/etc. indices.

The idea is taken from Qt (like models themselves), and the interface for persistent model indices is pretty much analogous to Qt's solution.

This PR implements the first part which is the `PersistentModelIndex` class itself. It behaves mostly like a `ModelIndex` does, but it also auto-registers itself with the source `Model` object when it is constructed. The `Model` will then return a `PersistentHandle` which the `PersistentModelIndex` will hold a weak reference to. This `PersistentHandle` object is owned by the `Model`, and will be updated (or deleted) when the model's data updates via the various events that the model can emit.

When a model has seen an update in its data that it can represent via inserts/moves/deletes, it will use the corresponding protected functions. These functions store a stack of operations delimited by matching `begin_*` and `end_*` calls for each operation. The implementation works this way due to some operations causing a cascade for later ones. For example, if multiple deletes of rows occur, the first actual deleted row must be the one at the bottom, since otherwise the indices for the other delete operations would be invalid and the model subclass would have to manually compensate for that. Instead, the model can just queue a bunch of `begin_delete_rows` operations and then execute them all by calling `end_delete_rows` a matching amount of times.

When an `end_*` call happens, the corresponding operation is popped from the operation stack and the `PersistentHandle` objects that are stored are modified. Since the handle objects are weakly pointed to by `PersistentModelIndex` objects, this means that the index that they are pointing to are automatically updated as well. Finally, after the operation is complete, a `did_update` is emitted to the clients to let them redraw, handle the updates etc. Alongside a generic `did_update`, more specific events are also fired based on operations. The reason that the generic `did_update` event is still fired is to not force applications which don't require granular operations to listen to granular events.

This PR also contains an example usage of this new granular updates API in `FileSystemModel`. `FileSystemModel` now handles events it obtains from `Core::FileWatcher` in order to pinpoint which file has changed/added/removed and only updates that file, using the system described above.

Please note that this PR doesn't as of yet bring any visible changes to the system due to most models that have updating data being wrapped in `SortingProxyModel`. `SortingProxyModel` does not currently handle these granular updates, and data is still lost even if the source model does a granular update. This will be fixed in the upcoming PRs by replacing `SortingProxyModel` and `FilteringProxyModel` to a common proxy model which handles both sorting and filtering and emits granular updates while doing so.

### CODE REVIEW

File: Userland/Libraries/LibGUI/FileSystemModel.cpp [hunk 1]
```diff
@@ -142,12 +138,29 @@ void FileSystemModel::Node::traverse_if_needed()
     }
 }
 
+OwnPtr<FileSystemModel::Node> FileSystemModel::Node::create_child(String const& child_name)
+{
+    String child_path = String::formatted("{}/{}", full_path(), child_name);
```
Comment (AtkinsSJ) [inline Userland/Libraries/LibGUI/FileSystemModel.cpp:143]: I think this wants to be `LexicalPath::join()`.

Comment (sin-ack) [reply Userland/Libraries/LibGUI/FileSystemModel.cpp:143]: Indeed. This code actually spawned `LexicalPath::join` but I guess all those changes were mangled together in the original PR.

File: Userland/Libraries/LibGUI/PersistentModelIndex.cpp [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+{
+    m_handle = other.m_handle;
+    return *this;
+}
+
+PersistentModelIndex& PersistentModelIndex::operator=(PersistentModelIndex&& other)
+{
+    m_handle = move(other.m_handle);
+    return *this;
+}
```
Comment (awesomekling) [inline Userland/Libraries/LibGUI/PersistentModelIndex.cpp:40]: Can these be `= default` instead?

**Comment by sin-ack** [comment]
One second, found some gnarly stuff in that last commit.

**Comment by awesomekling** [comment]
> This commit also makes a small change to SortingProxyModel which fixes a bug...

Can this be a separate commit?

File: Userland/Libraries/LibGUI/FileSystemModel.h [hunk 3]
```diff
@@ -81,9 +81,9 @@ class FileSystemModel
 
         FileSystemModel& m_model;
 
-        Node* parent { nullptr };
-        NonnullOwnPtrVector<Node> children;
-        bool has_traversed { false };
+        Node* m_parent { nullptr };
+        NonnullOwnPtrVector<Node> m_children;
+        bool m_has_traversed { false };
```
Comment (awesomekling) [inline Userland/Libraries/LibGUI/FileSystemModel.h:86]: The important changes would be a lot easier to read if we did these renames in a separate commit :)

**Comment by sin-ack** [comment]
> Can this be a separate commit?

Sure.

File: Userland/Libraries/LibGUI/FileSystemModel.h (hunk 3 above)
Comment (sin-ack) [reply Userland/Libraries/LibGUI/FileSystemModel.h:86]: Gotchu, will move the renames to a new commit.

**Pull request merged**